pkg/metrics/handler.go

```diff
@@ -1,25 +1,57 @@
 package metrics
 
 import (
+	"fmt"
 	"github.com/prometheus/client_golang/prometheus"
-	"go.uber.org/zap"
 )
 
 type MetricHandler interface {
 	Update(collector prometheus.Collector, value float64, labels ...string) error
 	Type() string
 }
 
-func (t *taskMetrics) UpdateMetric(name string, value float64, labels ...string) {
+type IncrementDecrementHandler interface {
+	MetricHandler
+	Inc(collector prometheus.Collector, labels ...string) error
+	Dec(collector prometheus.Collector, labels ...string) error
+}
+
+func (t *taskMetrics) performMetricAction(name string, action func(MetricHandler, prometheus.Collector, ...string) error, labels ...string) error {
 	t.mux.RLock()
 	metricDetail, ok := t.metrics[name]
 	t.mux.RUnlock()
+
 	if !ok {
-		zap.S().Errorf("Error: Metric not found for %s", name)
-		return
+		return fmt.Errorf("error: Metric not found for %s", name)
 	}
 
-	if err := metricDetail.Handler.Update(metricDetail.Collector, value, labels...); err != nil {
-		zap.S().Errorf("Error updating metric %s. Err: %s", name, err.Error())
+	if err := action(metricDetail.Handler, metricDetail.Collector, labels...); err != nil {
+		return fmt.Errorf("error performing action on metric %s. Err: %s", name, err.Error())
```

Comment:
why are we suing FMT instead of ZAP?

Reply:
I use it to build the error, I prefer to return the error and have the client be the one to log it.

Reply:
sounds reasonable